src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -883,41 +884,24 @@ DeferredErrorIfUnsupportedRecurringTuplesJoin(
 		}
 	}
 
-	if (recurType == RECURRING_TUPLES_REFERENCE_TABLE)
-	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "cannot pushdown the subquery",
-							 "There exist a reference table in the outer "
-							 "part of the outer join", NULL);
-	}
-	else if (recurType == RECURRING_TUPLES_FUNCTION)
-	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "cannot pushdown the subquery",
-							 "There exist a table function in the outer "
-							 "part of the outer join", NULL);
-	}
-	else if (recurType == RECURRING_TUPLES_EMPTY_JOIN_TREE)
-	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "cannot pushdown the subquery",
-							 "There exist a subquery without FROM in the outer "
-							 "part of the outer join", NULL);
-	}
-	else if (recurType == RECURRING_TUPLES_RESULT_FUNCTION)
-	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "cannot pushdown the subquery",
-							 "Complex subqueries, CTEs and local tables cannot be in "
-							 "the outer part of an outer join with a distributed table",
-							 NULL);
-	}
-	else if (recurType == RECURRING_TUPLES_VALUES)
+	/*
+	 * We recursively plan the outer joins where the outer side is a recurring
+	 * rel and the other side is a distributed one by wrapping the distributed
+	 * side into a subquery. However, we fail to do so if the distributed side is
+	 * a subquery that references the outer query --the only case that
+	 * RecursivelyPlanSubquery fails to plan a subquery.
+	 *
+	 * For this reason, we generate the error message based on this fact.
+	 */
+	if (recurType != RECURRING_TUPLES_INVALID)
 	{
+		char *errorDetail = psprintf("Outer joins are not supported when the "
+									 "inner side of the join is distributed "
+									 "and references a column from %s ",
```

Comment:
shouldn't we quote %s? The error messages are a bit confusing without that

Reply:
In general, this error message is too technical, users typically do not know inner side / outer side etc. 

Could we simplify this a bit, like `ERROR: outer joins with correlated subqueries are only supported when the correlation is on the distribution column`

(or anything better?)

Reply:
now we're using DeferredErrorIfUnsupportedLateralSubquery function to throw an error in that case, so this is outdated now

Reply:
this is totally dropped now